bcc/table.go

```diff
@@ -181,56 +185,134 @@ func (table *Table) DeleteAll() error {
 	return nil
 }
 
-// Iter returns a receiver channel to iterate over all table entries.
-func (table *Table) Iter() <-chan Entry {
+// TableIterator contains the current position for iteration over a *bcc.Table and provides methods for iteration.
+type TableIterator struct {
+	table *Table
+	fd    C.int
+
+	nextKey []byte
+	err     error
+
+	key     []byte
+	leaf    []byte
+	keyStr  []byte // lazily allocated
+	leafStr []byte // lazily allocated
+}
+
+// Iter returns an iterator to list all table entries available as raw bytes.
+func (table *Table) Iter() *TableIterator {
 	mod := table.module.p
-	ch := make(chan Entry, 128)
-	go func() {
-		defer close(ch)
-		fd := C.bpf_table_fd_id(mod, table.id)
-		key_size := C.bpf_table_key_size_id(mod, table.id)
-		leaf_size := C.bpf_table_leaf_size_id(mod, table.id)
-		key := make([]byte, key_size)
-		leaf := make([]byte, leaf_size)
-		keyP := unsafe.Pointer(&key[0])
-		leafP := unsafe.Pointer(&leaf[0])
-		alternateKeys := []byte{0xff, 0x55}
-		res := C.bpf_lookup_elem(fd, keyP, leafP)
-		// make sure the start iterator is an invalid key
-		for i := 0; i <= len(alternateKeys); i++ {
-			if res < 0 {
-				break
-			}
-			for j := range key {
-				key[j] = alternateKeys[i]
-			}
-			res = C.bpf_lookup_elem(fd, keyP, leafP)
-		}
-		if res == 0 {
-			return
-		}
-		keyStr := make([]byte, key_size*8)
-		leafStr := make([]byte, leaf_size*8)
-		keyStrP := (*C.char)(unsafe.Pointer(&keyStr[0]))
-		leafStrP := (*C.char)(unsafe.Pointer(&leafStr[0]))
-		for res = C.bpf_get_next_key(fd, keyP, keyP); res == 0; res = C.bpf_get_next_key(fd, keyP, keyP) {
-			r := C.bpf_lookup_elem(fd, keyP, leafP)
-			if r != 0 {
-				continue
-			}
-			r = C.bpf_table_key_snprintf(mod, table.id, keyStrP, C.size_t(len(keyStr)), keyP)
-			if r != 0 {
-				break
-			}
-			r = C.bpf_table_leaf_snprintf(mod, table.id, leafStrP, C.size_t(len(leafStr)), leafP)
-			if r != 0 {
-				break
-			}
-			ch <- Entry{
-				Key:   string(keyStr[:bytes.IndexByte(keyStr, 0)]),
-				Value: string(leafStr[:bytes.IndexByte(leafStr, 0)]),
-			}
+	fd := C.bpf_table_fd_id(mod, table.id)
+	keySize := C.bpf_table_key_size_id(mod, table.id)
+	leafSize := C.bpf_table_leaf_size_id(mod, table.id)
+	key := make([]byte, keySize)
+	leaf := make([]byte, leafSize)
+
+	nextKey := make([]byte, keySize)
+	keyP := unsafe.Pointer(&nextKey[0])
+
+	res := C.bpf_get_first_key(fd, keyP, keySize)
+	if res != 0 {
+		return &TableIterator{
+			err: errNoIterationStartingPoint,
 		}
-	}()
-	return ch
+	}
+
+	return &TableIterator{
+		table:   table,
+		fd:      fd,
+		nextKey: nextKey,
+
+		key:  key,
+		leaf: leaf,
+	}
+}
+
+// Next looks up the next element and return true if one is available.
+func (it *TableIterator) Next() bool {
+	if it.err != nil || it.nextKey == nil {
+		return false
+	}
+
+	it.key, it.nextKey = it.nextKey, it.key
+
+	keyP := unsafe.Pointer(&it.key[0])
+	leafP := unsafe.Pointer(&it.leaf[0])
+	res := C.bpf_lookup_elem(it.fd, keyP, leafP)
```

Comment:
Again with `bufio.Scanner` in mind: what do you think about e.g.

```
diff --git a/bcc/table.go b/bcc/table.go
index a8c1954..b304beb 100644
--- a/bcc/table.go
+++ b/bcc/table.go
@@ -18,6 +18,7 @@ import (
        "bytes"
        "errors"
        "fmt"
+       "os"
        "unsafe"
 )
 
@@ -238,16 +239,22 @@ func (it *TableIterator) Next() bool {
 
        keyP := unsafe.Pointer(&it.key[0])
        leafP := unsafe.Pointer(&it.leaf[0])
-       res := C.bpf_lookup_elem(it.fd, keyP, leafP)
+       res, err := C.bpf_lookup_elem(it.fd, keyP, leafP)
        if res != 0 {
-               it.err = errIterationFailed
+               if !os.IsNotExist(err) {
+                       it.err = err
+               }
                return false
        }
 
        nextKeyP := unsafe.Pointer(&it.nextKey[0])
-       res = C.bpf_get_next_key(it.fd, keyP, nextKeyP)
+       res, err = C.bpf_get_next_key(it.fd, keyP, nextKeyP)
        if res != 0 {
                it.nextKey = nil
+               if !os.IsNotExist(err) {
+                       it.err = err
+                       return false
+               }
        }
 
        return true
```

I.e.

* when `Next()` returns `false` and `Err()` returns `== nil`, the end of the table is reached,
* when `Next()` returns `false` and `Err()` returns `!= nil`, the returned error occured.

Reply:
We could then also check that `iter.Err()` is `nil` in the tests.